### PR TITLE
Re-try addition of configurable trace sampling strategy

### PR DESCRIPTION
**Pull request by cboggs**
EDIT: Please see #703 for description

@JML - I'm not entirely sure that adding an override to Gopkg.toml was the proper fix, so if I got it wrong, let me know what the proper method is and I'll fix it up. :-)

Thanks!

### CODE REVIEW

**Comment by jml** [comment]
Thanks! Will get around to this next week.

No idea about the dep stuff, but at a guess you don't need the constraint,
you just need to run `dep update` (or `upgrade`, I forget exact spelling).

**Comment by cboggs** [comment]
Gotcha. I didn't do the `dep ensure-update` or whatever it was, just because it implied that it would update ALL dependencies, and that made me nervous. I'll give it a shot in a bit and see what the damage is. :-)

**Comment by bboreham** [comment]
Just run `dep ensure -update <path>` to update one package.

**Comment by bboreham** [comment]
Also you must check in `Gopkg.lock` when you change the vendor'd libs.

**Comment by bboreham** [comment]
Since there are a couple more updates brought in from `weaveworks/common`, I think it is worth calling them out:

Log gRPC request on error weaveworks/common#85
Don't log request body and headers for 502 errors weaveworks/common#84

And now I'm a bit worried by weaveworks/common#85 - will it log the entire set of samples on a failed `Push` ?

**Comment by cboggs** [comment]
@bboreham, thanks for the pointers! I've removed the override, but the vendor update should still be in place.

I'm not sure how to go about answering your question re: https://github.com/weaveworks/common/pull/85, unfortunately. Any guidance there?

**Comment by awh** [comment]
> will it long the entire set of samples on a failed push?

**Comment by jml** [comment]
@bboreham Do you object to us merging this as is? What do we need to do to get it ready to merge?

**Comment by bboreham** [comment]
I would suggest testing the error path and then deciding if that level of logging is acceptable.

**Comment by bboreham** [comment]
I notice #705 also contains the update to latest version of dep, so one of them is going to have to rebase after merge of the other.

**Comment by cboggs** [comment]
I'd be glad to wait for #705 @bboreham and @tomwilkie. The `dep` speedup will be very nice to have, as it's been painful keeping up the vendor bits in this branch. :-)

**Comment by bboreham** [comment]
Has the error logging been tested yet?

**Comment by cboggs** [comment]
Finally just blew away my local branch and forced things to a sane state from latest master.

@bboreham I'm not sure how to go about testing the error logging, short of letting it run for a while in our staging cluster and see what happens in the logs.

Do you know of a quick / easy-ish way to cause this failure pattern so that I can give some quicker feedback?

Thanks!

**Comment by cboggs** [comment]
@bboreham I think I may have inadvertently found an instance of what you expect to see in the logs with the recent change to common:
```
time="2018-02-27T23:03:19Z" level=warning msg=gRPC duration=8.160544586s error="rpc error: code = Code(400) desc = sample timestamp out of order for series <series bits snipped>; last timestamp: 1519772598.344, incoming t
imestamp: 1519772588.345" method=/cortex.Ingester/Push request="&WriteRequest{Timeseries:[{[{[95 95 110 97 109 101 95 95] [102 116 95 104 119 95 102 111 114 101 99 97 115 116 58 102 116 95 97 103 103 114 101 103 97 116 105 111 110 58 102... <many many tens of thousands more bytes>
```

Sure looks like a log of all samples in the failed push. Let me know if that's not what you were expecting and I'll dig a bit more.

**Comment by bboreham** [comment]
Thanks, yes, that is what I was wo

**Comment by bboreham** [comment]
Sorry, phone going haywire. Will come back tomorrow.

**Comment by bboreham** [comment]
Tuncating may be a workable compromise; however that dump you gave as an example doesn't look readable at any length. `95 95 110 97 109 101 95 95` is "__name__" - does the dumping code look for a `Stringer` we could add?

**Comment by cboggs** [comment]
Definitely not readable at any length. I'll try to peruse the dumping code a bit next week once we've tackled a few other things, and see if we can at least make it readable for some preamble length.

**Comment by cboggs** [comment]
@bboreham, @jml, do you think the above-referenced PR will address this sufficiently?

If so, once it's merged I'll vendor it into this branch and add the config option to the ingester, and we should be golden.

**Comment by jml** [comment]
I abstain. Sorry!

**Comment by bboreham** [comment]
PR in common looks plausible; it's not clear what the answer to my `Stringer` question is. I would much rather see `"__name__"` than 95 `95 110 97 109 101 95 95`

**Comment by cboggs** [comment]
@bboreham, I had a quick clarifying question for you on the logging output issue.

Is there a particular reason you'd like to see the string-ified version of the byte array that's being logged? I ask because the full series is being logged prior to the byte array, and the only additional information to be seen in the request body is the full set of samples and timestamps. String-ifying this would indeed shrink the output by a small bit (more accurately, by the number of bytes that make up the delta between each char in the series string and their corresponding ASCII value lengths, plus whitespace), but the rest of the output will be a big nested array of floats.

If this is definitely needed, I can spend the time to figure out a way to do so (it's still not proving to be as trivial as I'd hoped), but I'd like to better understand that need first. :-)

Thanks!

**Comment by bboreham** [comment]
I don't specially need to see the string; I just think it's dumb to print something we know is a string as the ascii codes.

Happy to have that filed as a clean-up issue afterwards.

**Comment by cboggs** [comment]
Agreed that it's dumb to print a string as raw ascii codes. :-) I just think it's easier and (in this specific case) pretty reasonable to just omit that massive output entirely.

**Comment by bboreham** [comment]
I went back to https://github.com/weaveworks/common/pull/89; I can't see where it is truncating at N bytes - it seems to be just omitting the request details completely.

**Comment by cboggs** [comment]
Yes, that's true. Maybe "truncate" is the wrong word.

I started down the path of a byte limit, but given that the data at the beginning of the ingester push requests is duplicate data from earlier in the log message, it seemed reasonable to add a configurable omission of the request body.

**Comment by cboggs** [comment]
Hmm... Maybe that is indeed too sledgehammer-y. I forget that ingesters take more than just `Push` requests.

I'll try the byte limiter again, and retain the configuration option for such. Stand by!

**Comment by cboggs** [comment]
@bboreham, does this PR seem OK now with the update to [weaveworks/common](https://github.com/weaveworks/common/pull/89)?

File: cmd/querier/main.go [hunk 1]
```diff
@@ -46,8 +46,8 @@ func main() {
 	flag.Parse()
 
 	// Setting the environment variable JAEGER_AGENT_HOST enables tracing
-	jaegerAgentHost := os.Getenv("JAEGER_AGENT_HOST")
-	trace := tracing.New(jaegerAgentHost, "querier")
+	trace := tracing.NewFromEnv("querier")
+	defer trace.Close()
```
Comment (csmarchbanks) [inline cmd/querier/main.go]: don't need two defers

File: cmd/ingester/main.go [hunk 2]
```diff
@@ -26,6 +27,7 @@ func main() {
 			GRPCMiddleware: []grpc.UnaryServerInterceptor{
 				middleware.ServerUserHeaderInterceptor,
 			},
+			ExcludeRequestInLog: false,
```
Comment (bboreham) [inline cmd/ingester/main.go]: I think you wanted this `true`

**Comment by cboggs** [comment]
Whoo! Thanks a million @bboreham for your help (and patience)!

**Pull request merged**